postgres-operator-sdk/README.md

```diff
@@ -0,0 +1,288 @@
+      The handler needs to know where minikube is and how to execute functions to interact with 
+      minikube. The above path is the default path to the .kube/config. If for whatever reason
+      your path is different, make the necessary changes to match your path.
+
+Open a new terminal while the Operator is running in the other one.
+
+Run the following command to begin testing:
+
+```
+$ kubectl create -f deploy/client1.yaml
+```
+
+The above command creates a Postgres instance named client1 that contains a Postgres image 
+and several Postgres queries to be ran upon launch of the instance.
+The Postgres Operator should catch this deployment and begin creating the corresponding Deployment and Service objects.
+
+Run the following commands to verify that the operator is doing what it is supposed to do:
+
+```
+$ kubectl get crd
+$ kubectl get postgres client1
+$ kubectl describe postgres client1
+$ psql -h <IP> -p <port> -U <username> -d <db-name>
+```
+For the last command above, plug in the values of the IP and port numbers generated in the Operator output
+as per service object availability.
+For the username and db-name, go to deploy/client1.yaml and use the values for the username and database specified.
+When prompted for password, use the password in the client1.yaml as well.
+
+## Acknowledgements
+This Postgres operator is completely based off of Cloud-Ark's [Postgres Custom Controller](https://github.com/cloud-ark/kubeplus/tree/master/postgres-crd-v2) as is most of the code in the Handler.go file
+
+------
+
+## Purpose of Experiment
+
+--> Talk about the hypothesis of the experiment that Operator SDK might be less performant than sample-controller.
+Give the reason why we think this might be the case. Refer to the Github issue about this from Operator SDK repo.
+Refer to CloudARK blog on 'Writing Kubernetes Custom Controller'
+to build your understanding of Listers, Informers, etc.
+Then present your understanding of what a Lister is and why you think Operator SDK might be less performant.
+Point out specific places in the Operator SDK code where you see the problem.
+
+After reading the Operator SDK library and the different CRUD functions inplemented in the sdk file, it became apparent that
+the SDK library does not use Listers upon performing CRUD operations and instead uses dynamic clientsets.
+
+Listers play a role in serving as objects that allow indexing into a client side cache for Kubernetes resources. Rather than write to
+the server upon every call performed, the Lister allows for changes and retrievals to be made on the client side in which changes to the
+resource are later synced with the server.
+
+For an overview of the Custom Controller architecture, refer to the [Cloud-ARK](https://medium.com/@cloudark/kubernetes-custom-controllers-b6c7d0668fdf) blog post.
+
+[Issues](https://github.com/operator-framework/operator-sdk/issues/258) related to the usage of Clientsets in Operator SDK.
```

Comment:
This issue has been closed now. Looks Operator SDK is getting re-written to use controller-runtime libraries. Check the details here:
https://github.com/operator-framework/operator-sdk/issues/258

This will probably mean that the new version of Operator SDK might not have any performance problems related to not using Listers.

Take a look at the controller-runtime libraries and following issue

https://github.com/operator-framework/operator-sdk/issues/382.

In the issue description they have mentioned a branch where the new work is happening. That branch does not seem to exist anymore. So probably the work is now merged in their master. If so, using latest Operator SDK might work differently than the version that you used. So make sure that you save your version of Operator SDK as you will need it when running experiments and when documenting the results.

Also, we might want to change our experiments a bit. Instead of implementing another version of Postgres which uses Listers, as we had discussed earlier, more worthwhile will be to implement Postgres with the version of Operator SDK that uses controller-runtime libraries.

Lets discuss this in today's meeting. Check out the controller-runtime libraries